File: src/C/extract_matrix.c
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-20 15:41:14
@LastEditors: HoodUSSEnterprise
@LastEditTime: 2026-06-24 13:32:38
@FilePath: \asm_matrix_benchmark\src\C\extract_matrix.c
@Description: extract matrix c code
*************************************************************/

#include "extract_matrix.h"

/***********************************************************
@description: extract matrix's row
@param {MatrixInt} *m
@param {size_t} index
@return {*}
 ************************************************************/
MatrixInt *extract_row_int(MatrixInt *m, size_t index)
{
    // check m and m->data
    if (m == NULL || m->data == NULL)
    {
        fprintf(stderr, "Invalid param!\n");
        return NULL;
    }
    // check index, index must in range 0 and m->rows - 1
    if (index > m->rows - 1)
    {
        fprintf(stderr, "Invalid param!\n");
        return NULL;
    }
    // malloc for new res
    MatrixInt *res = NULL;
    res = (MatrixInt *)malloc(sizeof(MatrixInt));
    if (res == NULL)
    {
        fprintf(stderr, "Memory allocation failed\n");
        return NULL;
    }
    res->cols = m->cols;
    res->rows = 1;
    // malloc for res->data
    res->data = (int *)malloc(sizeof(int) * res->cols);
    if (res->data == NULL)
    {
        free(res);
        fprintf(stderr, "Memory allocation failed\n");
        return NULL;
    }
    for (size_t i = 0; i < m->cols; i++)
    {
        res->data[i] = m->data[index * m->cols + i];
    }
    return res;
}

/***********************************************************
@description: extract matrix's col
@param {MatrixInt} *m
@param {size_t} index
@return {*}
 ************************************************************/
MatrixInt *extract_col_int(MatrixInt *m, size_t index)
{
    // check m and m->data
    if (m == NULL || m->data == NULL)
    {
        fprintf(stderr, "Invalid param!\n");
        return NULL;
    }
    // check index, index must in range 0 and m->cols - 1
    if (index > m->cols - 1)
    {
        fprintf(stderr, "Invalid param!\n");
        return NULL;
    }
    // malloc for new res
    MatrixInt *res = NULL;
    res = (MatrixInt *)malloc(sizeof(MatrixInt));
    if (res == NULL)
    {
        fprintf(stderr, "Memory allocation failed\n");
        return NULL;
    }
    res->rows = m->rows;
    res->cols = 1;
    // malloc for res->data
    res->data = (int *)malloc(sizeof(int) * res->rows);
    if (res->data == NULL)
    {
        free(res);
        fprintf(stderr, "Memory allocation failed\n");
        return NULL;
    }
    for (size_t i = 0; i < m->rows; i++)
    {
        res->data[i] = m->data[i * m->cols + index];
    }
    return res;
}

/***********************************************************
@description: extract matrix's diag, that means the number which row index equals to col index
@param {MatrixInt} *m
@return {*}
 ************************************************************/
int *extract_diag_int(MatrixInt *m)
{
    // check m and m->data
    if (m == NULL || m->data == NULL)
    {
        fprintf(stderr, "Invalid param!\n");
        return NULL;
    }
    size_t data_len = (m->rows >= m->cols ? m->cols : m->rows);
    int *data = (int *)malloc(sizeof(int) * data_len);
    if (data == NULL)
    {
        fprintf(stderr, "Memory allocation failed\n");
        return NULL;
    }
    for (size_t i = 0; i < data_len; i++)
    {
        data[i] = m->data[i * m->cols + i];
    }
    return data;
}

/***********************************************************
@description: extract matrix's row double
@param {MatrixDouble} *m
@param {size_t} index
@return {*}
************************************************************/
MatrixDouble *extract_row_double(MatrixDouble *m, size_t index)
{
    // check m and m->data
    if (m == NULL || m->data == NULL)
    {
        fprintf(stderr, "Invalid param!\n");
        return NULL;
    }
    // check index, index must in range 0 and m->rows - 1
    if (index > m->rows - 1)
    {
        fprintf(stderr, "Invalid param!\n");
        return NULL;
    }
    // malloc for new res
    MatrixDouble *res = NULL;
    res = (MatrixDouble *)malloc(sizeof(MatrixDouble));
    if (res == NULL)
    {
        fprintf(stderr, "Memory allocation failed\n");
        return NULL;
    }
    res->cols = m->cols;
    res->rows = 1;
    // malloc for res->data
    res->data = (double *)malloc(sizeof(double) * res->cols);
    if (res->data == NULL)
    {
        free(res);
        fprintf(stderr, "Memory allocation failed\n");
        return NULL;
    }
    for (size_t i = 0; i < m->cols; i++)
    {
        res->data[i] = m->data[index * m->cols + i];
    }
    return res;
}

/***********************************************************
@description: extract matrix's col double
@param {MatrixDouble} *m
@param {size_t} index
@return {*}
************************************************************/
MatrixDouble *extract_col_double(MatrixDouble *m, size_t index)
{
    // check m and m->data
    if (m == NULL || m->data == NULL)
    {
        fprintf(stderr, "Invalid param!\n");
        return NULL;
    }
    // check index, index must in range 0 and m->cols - 1
    if (index > m->cols - 1)
    {
        fprintf(stderr, "Invalid param!\n");
        return NULL;
    }
    // malloc for new res
    MatrixDouble *res = NULL;
    res = (MatrixDouble *)malloc(sizeof(MatrixDouble));
    if (res == NULL)
    {
        fprintf(stderr, "Memory allocation failed\n");
        return NULL;
    }
    res->rows = m->rows;
    res->cols = 1;
    // malloc for res->data
    res->data = (double *)malloc(sizeof(double) * res->rows);
    if (res->data == NULL)
    {
        free(res);
        fprintf(stderr, "Memory allocation failed\n");
        return NULL;
    }
    for (size_t i = 0; i < m->rows; i++)
    {
        res->data[i] = m->data[i * m->cols + index];
    }
    return res;
}

/***********************************************************
@description: extract matrix's diag double
@param {MatrixDouble} *m
@return {*}
************************************************************/
double *extract_diag_double(MatrixDouble *m)
{
    // check m and m->data
    if (m == NULL || m->data == NULL)
    {
        fprintf(stderr, "Invalid param!\n");
        return NULL;
    }
    size_t data_len = (m->rows >= m->cols ? m->cols : m->rows);
    double *data = (double *)malloc(sizeof(double) * data_len);
    if (data == NULL)
    {
        fprintf(stderr, "Memory allocation failed\n");
        return NULL;
    }
    for (size_t i = 0; i < data_len; i++)
    {
        data[i] = m->data[i * m->cols + i];
    }
    return data;
}

File: src/JavaScript/extract_matrix.js
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-29 20:10:32
@LastEditors: GitHubCopilot
@LastEditTime: 2026-06-29 20:10:32
@FilePath: \asm_matrix_benchmark\src\JavaScript\extract_matrix.js
@Description: extract matrix JavaScript code
*************************************************************/

"use strict";

const { checkMatrix, buildTypedMatrix } = require("./common");

/**
 * Extract one row as a 1 x cols matrix.
 * @param {{rows:number, cols:number, data:number[]}} m
 * @param {number} index
 * @param {"int"|"float"|"double"} type
 * @returns {{rows:number, cols:number, data:number[]}|null}
 */
function extractRow(m, index, type) {
  if (!checkMatrix(m, "m") || !Number.isInteger(index)) {
    console.error("Invalid param!");
    return null;
  }
  if (index < 0 || index >= m.rows) {
    console.error("Index out of range!");
    return null;
  }

  const data = new Array(m.cols);
  for (let j = 0; j < m.cols; j++) {
    data[j] = m.data[index * m.cols + j];
  }
  return buildTypedMatrix(1, m.cols, data, type);
}

/**
 * Extract one column as a rows x 1 matrix.
 * @param {{rows:number, cols:number, data:number[]}} m
 * @param {number} index
 * @param {"int"|"float"|"double"} type
 * @returns {{rows:number, cols:number, data:number[]}|null}
 */
function extractCol(m, index, type) {
  if (!checkMatrix(m, "m") || !Number.isInteger(index)) {
    console.error("Invalid param!");
    return null;
  }
  if (index < 0 || index >= m.cols) {
    console.error("Index out of range!");
    return null;
  }

  const data = new Array(m.rows);
  for (let i = 0; i < m.rows; i++) {
    data[i] = m.data[i * m.cols + index];
  }
  return buildTypedMatrix(m.rows, 1, data, type);
}

/**
 * Extract main diagonal values as a plain array.
 * @param {{rows:number, cols:number, data:number[]}} m
 * @returns {number[]|null}
 */
function extractDiag(m) {
  if (!checkMatrix(m, "m")) {
    return null;
  }

  const n = Math.min(m.rows, m.cols);
  const diag = new Array(n);
  for (let i = 0; i < n; i++) {
    diag[i] = m.data[i * m.cols + i];
  }
  return diag;
}

function extract_row_int(m, index) {
  return extractRow(m, index, "int");
}

function extract_col_int(m, index) {
  return extractCol(m, index, "int");
}

function extract_diag_int(m) {
  return extractDiag(m);
}

function extract_row_float(m, index) {
  return extractRow(m, index, "float");
}

function extract_col_float(m, index) {
  return extractCol(m, index, "float");
}

function extract_diag_float(m) {
  return extractDiag(m);
}

function extract_row_double(m, index) {
  return extractRow(m, index, "double");
}

function extract_col_double(m, index) {
  return extractCol(m, index, "double");
}

function extract_diag_double(m) {
  return extractDiag(m);
}

module.exports = {
  extract_row_int,
  extract_col_int,
  extract_diag_int,
  extract_row_float,
  extract_col_float,
  extract_diag_float,
  extract_row_double,
  extract_col_double,
  extract_diag_double,
};
